crates/recursor/src/recursor_dns_handle.rs

```diff
@@ -226,6 +262,77 @@ impl RecursorDnsHandle {
         }
     }
 
+    /// Handle CNAME expansion for the current query
+    #[async_recursion]
+    async fn resolve_cnames(
+        &self,
+        mut lookup: Lookup,
+        query: Query,
+        now: Instant,
+        query_has_dnssec_ok: bool,
+        mut depth: u8,
+    ) -> Result<Lookup, Error> {
+        let query_type = query.query_type();
+        let query_name = query.name().clone();
+
+        self.record_cache
+            .insert_records(query, lookup.records().iter().map(Record::to_owned), now);
+
+        // Don't resolve CNAME lookups for a CNAME (or ANY) query
+        if query_type == RecordType::CNAME || query_type == RecordType::ANY {
+            return Ok(lookup);
+        }
+
+        depth += 1;
+        if self.recursion_limit > 0 && depth > self.recursion_limit {
+            warn!("Recursion depth exceeded for {query_name:?}");
+            return Err(ErrorKind::Proto(
+                ProtoErrorKind::NotAllRecordsWritten {
+                    count: depth as usize,
+                }
+                .into(),
+            )
+            .into());
+        }
+
+        let mut cname_chain = vec![];
+        for rec in lookup.records().iter() {
+            if let CNAME(name) = rec.data() {
+                let cname_query = Query::query(name.0.clone(), query_type);
+
+                // Note that we aren't worried about whether the intermediates are local or remote
+                // to the original queried name, or included or not included in the original
+                // response.  Resolve will either pull the intermediates out of the cache or query
+                // the appropriate nameservers if necessary.
+                let records = match self
+                    .resolve(cname_query, now, query_has_dnssec_ok, depth)
+                    .await
+                {
+                    Ok(cname_r) => cname_r,
+                    Err(e) => {
+                        return Err(e);
+                    }
+                };
+
+                // Here, we're looking for either the terminal record type (matching the
+                // original query, or another CNAME.
+                cname_chain.extend(records.records().iter().filter_map(|r| {
+                    if r.record_type() == query_type || r.record_type() == RecordType::CNAME {
+                        Some(Record::to_owned(r))
+                    } else {
+                        None
+                    }
```

Comment:
Maybe like this?

```rust
match r.record_type() {
    ty @ RecordType::CNAME if ty == query_type => Some(r.to_owned()),
    _ => None,
}
```

Reply:
That match arm is effectively r.record_type() == RecordType::CNAME && r.record_type == query_type, whereas the original is looking for CNAME || the original query type.

I tried a few match variations to recreate the || logic, but at least what I came up with read less clearly than the original if statement.